### PR TITLE
Update IsFailure method to reflect Run retries

### PR DESCRIPTION
**Pull request by lbernick**
# Changes
Custom Tasks support retries (as of #4327); however, there are some places in the code base that consider
retries only for TaskRuns. This commit updates ResolvedPipelineRunTask.IsFailure to recognize
Run retries and adds tests for this method. The function PipelineRunState.getNextTasks will
be updated in a future commit. No functional changes for TaskRuns.

/kind bug

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

```release-note
[Bug Fix]: Runs that have remaining retries are not failed
```

### CODE REVIEW

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 1]
```diff
@@ -136,16 +136,46 @@ func (t ResolvedPipelineRunTask) IsSuccessful() bool {
 
 // IsFailure returns true only if the run has failed and will not be retried.
 func (t ResolvedPipelineRunTask) IsFailure() bool {
-	if t.IsCustomTask() {
-		return t.Run != nil && t.Run.IsDone() && !t.Run.IsSuccessful()
+	if t.IsCancelled() {
```
Comment (lbernick) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:139]: one weird bit of maybe undefined behavior is that for TaskRuns, if it is marked as cancelled but has status "unknown" (rather than "false"), this is considered not failed and not done. I decided to use the same behavior for Runs as for TaskRuns. If we want to change this behavior I think it belongs in a separate commit.

**Review by vdemeester — APPROVED** [review]

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go (hunk 1 above)
Comment (pritidesai) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:139]: A taskRun is considered as Cancelled if `status` is `false` and `reason` is set to `TaskRunCancelled`. Are you saying the `status` is set to `unknown` when a taskRun is cancelled on demand? I didnt quite follow how thats happening.

**Comment by pritidesai** [comment]
thank you @lbernick 👍 
/lgtm

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go (hunk 1 above)
Comment (lbernick) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:139]: What I mean is, if a TaskRun has status "unknown" with reason "cancelled", it will be marked as not failed. It's unclear to me whether such a condition can occur and what the reconciler's behavior should be in this scenario. Our tests implicitly depended on this behavior prior to this commit, and now make this behavior explicit.
